inox2d/src/render.rs

```diff
@@ -1,378 +1,386 @@
+						});
+
+						swap(
+							&mut zsorted_children_list,
+							&mut comps
+								.get_mut::<CompositeRenderCtx>(node.uuid)
+								.unwrap()
+								.zsorted_children_list,
+						);
+					}
+					// for TexturedMesh, obtain and write deforms into vertex_buffer
+					DrawableKind::TexturedMesh(..) => {
+						let render_ctx = comps.get::<TexturedMeshRenderCtx>(node.uuid).unwrap();
+						let vert_offset = render_ctx.vert_offset as usize;
+						let vert_len = render_ctx.vert_len;
+						let deform_stack = comps
+							.get::<DeformStack>(node.uuid)
+							.expect("A TexturedMesh must have an associated DeformStack.");
+
+						deform_stack.combine(
+							nodes,
+							comps,
+							&mut self.vertex_buffers.deforms[vert_offset..(vert_offset + vert_len)],
+						);
+					}
+				}
 			}
 		}
 
-		Self {
-			vertex_buffers,
-			root_drawables_zsorted,
-			node_render_ctxs,
-		}
+		root_drawable_uuid_zsort_vec.sort_by(|a, b| a.1.total_cmp(&b.1).reverse());
+		self.root_drawables_zsorted
+			.iter_mut()
+			.zip(root_drawable_uuid_zsort_vec.iter())
+			.for_each(|(old, new)| *old = new.0);
 	}
-}
 
-impl Puppet {
-	/// Update the puppet's nodes' absolute transforms, by combining transforms
-	/// from each node's ancestors in a pre-order traversal manner.
-	pub fn update_trans(&mut self) {
-		let root_node = self.nodes.arena[self.nodes.root].get();
-		let node_rctxs = &mut self.render_ctx.node_render_ctxs;
-
-		// The root's absolute transform is its relative transform.
-		let root_trans = node_rctxs.get(&root_node.uuid).unwrap().trans_offset.to_matrix();
-
-		// Pre-order traversal, just the order to ensure that parents are accessed earlier than children
-		// Skip the root
-		for id in self.nodes.root.descendants(&self.nodes.arena).skip(1) {
-			let node_index = &self.nodes.arena[id];
-			let node = node_index.get();
-
-			if node.lock_to_root {
-				let node_render_ctx = node_rctxs.get_mut(&node.uuid).unwrap();
-				node_render_ctx.trans = root_trans * node_render_ctx.trans_offset.to_matrix();
-			} else {
-				let parent = &self.nodes.arena[node_index.parent().unwrap()].get();
-				let parent_trans = node_rctxs.get(&parent.uuid).unwrap().trans;
-
-				let node_render_ctx = node_rctxs.get_mut(&node.uuid).unwrap();
-				node_render_ctx.trans = parent_trans * node_render_ctx.trans_offset.to_matrix();
-			}
-		}
+	/// Memory layout: `[[x, y], [x, y], ...]`
+	pub fn get_raw_verts(&self) -> &[f32] {
+		VertexBuffers::vec_vec2_as_vec_f32(&self.vertex_buffers.verts)
+	}
+	/// Memory layout: `[[x, y], [x, y], ...]`
+	pub fn get_raw_uvs(&self) -> &[f32] {
+		VertexBuffers::vec_vec2_as_vec_f32(&self.vertex_buffers.uvs)
+	}
+	/// Memory layout: `[[i0, i1, i2], [i0, i1, i2], ...]`
+	pub fn get_raw_indices(&self) -> &[u16] {
+		self.vertex_buffers.indices.as_slice()
+	}
+	/// Memory layout: `[[dx, dy], [dx, dy], ...]`
+	pub fn get_raw_deforms(&self) -> &[f32] {
+		VertexBuffers::vec_vec2_as_vec_f32(&self.vertex_buffers.deforms)
 	}
```

Comment:
These getters aren't necessary. They add noise more than anything right now. Better just use the code that's inside them on the one call site where they're used and keep every field of the `RenderCtx` as public.

Reply:
Now I think it, such interfaces are loosely inspired by the "slab allocator" talk by Luna, where node data for GPU can be efficiently re-ordered for efficient batched rendering. In such case, actual data layout will be complex, but renderer implementers still want to see a contiguous memory region for upload. And this is where such "adapter" functions come in.

But true, we are so far away from actually having an idea of how to design such an allocator, so probably a good idea to remove these getters for now. We will have a better idea of what the interface should be when internal memory layouts really begin to change.